File: 3 - Introduce Polymorphic Creation with Factory Method/JS/src/DomBuilder.js
export default class DomBuilder {
  constructor (root) {
    this.structure = {}
    this.structure[root] = {}
  }

  addAbove (element) {
    const newStructure = {}
    newStructure[element] = this.structure
    this.structure = newStructure
  }

  generate () {
    const elements = Object.keys(this.structure).map(e => {
      const subElements = Object.keys(this.structure[e]).map(subE => `<${subE}></${subE}>`)
      return `<${e}>${subElements.join('')}</${e}>`
    })

    return elements.join('')
  }
}

File: 3 - Introduce Polymorphic Creation with Factory Method/C++/before/DomBuilder.cpp
#include "DomBuilder.h"

#include <string>

DomBuilder::DomBuilder(std::string root) : Root(root) {};

void DomBuilder::AddBelow(std::string tag)
{
}

void DomBuilder::AddAbove(std::string tag)
{
  throw 0;
}
